runtime/src/bank.rs

```diff
@@ -5044,20 +5044,42 @@ impl Bank {
                 .is_active(&feature_set::consistent_recent_blockhashes_sysvar::id()),
         }
     }
+
+    /// Bank cleanup
+    ///
+    /// If the bank is unfrozen and then dropped, additional cleanup is needed.  In particular,
+    /// cleaning up the pubkeys that are only in this bank.  To do that, call into AccountsDb to
+    /// scan for dirty pubkeys and add them to the uncleaned pubkeys list so they will be cleaned
+    /// up in AccountsDb::clean_accounts().
+    fn cleanup(&self) {
+        if self.is_frozen() {
+            // nothing to do here
+            return;
+        }
+
+        self.rc
+            .accounts
+            .accounts_db
+            .scan_slot_and_insert_dirty_pubkeys_into_uncleaned_pubkeys(self.slot);
+    }
 }
 
 impl Drop for Bank {
     fn drop(&mut self) {
-        if !self.skip_drop.load(Relaxed) {
-            if let Some(drop_callback) = self.drop_callback.read().unwrap().0.as_ref() {
-                drop_callback.callback(self);
-            } else {
-                // Default case
-                // 1. Tests
-                // 2. At startup when replaying blockstore and there's no
-                // AccountsBackgroundService to perform cleanups yet.
-                self.rc.accounts.purge_slot(self.slot());
-            }
+        if self.skip_drop.load(Relaxed) {
+            return;
+        }
+
+        self.cleanup();
```

Comment:
nice! I think this is the right place to put this in relation to the `Some(drop_callback)` check below, subtle!

Reply:
Yeah, I was wondering about where exactly to call `cleanup()`. Since the drop triggers `clean_accounts()`, and `uncleaned_pubkeys` needs to have all the pubkeys beforehand, it seems like `cleanup()` needs to be called first.

Reply:
yes also, regardless of whether there is a `drop_callback()`, those dirty pubkeys are still in storage, so they need to be added to the list!